### Fine-tuning the model (DataCamp)

In [89]:
# Confustion matrix
'''                         Predicted:            Predicted:
                            Positive              Negative

Actual: Positive           True Positive(TP)      False Negative(FN)
Actual: Negative           False Positive(FP)     True Negative(TN)'''

'                         Predicted:            Predicted:\n                            Positive              Negative\n\nActual: Positive           True Positive(TP)      False Negative(FN)\nActual: Negative           False Positive(FP)     True Negative(TN)'

Precision:
TRUE POSITIVES/ (TRUE POSITIVES + FALSE POSITIVES)
- High precision = lower false positive rates

Recall:
TRUE POSITIVES/ (TRUE POSITIVES + FALSE NEGATIVES)
- High recall = lower false negative rate

F1 score:
2*(precision * recall)/ precision + recall
- High F1 score = high F1 score, high precision and recall (∈ [0,1])

In [90]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix, precision_score, f1_score
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt

In [91]:
NAC_data =  pd.read_csv('NAC_data.csv')
NAC_df = pd.DataFrame(NAC_data)
NAC_df.head()

Unnamed: 0          Player              Team  \
0           0     P. Iemmello         Catanzaro   
1           1  J. Petriccione           Crotone   
2           2       T. Biasci         Catanzaro   
3           3   E. Volpicelli  Sangiuliano City   
4           4       A. Curcio         Catanzaro   

  Team within selected timeframe         Position   Age  Market value  \
0                      Catanzaro               CF  31.0        750000   
1                        Crotone  DMF, RDMF, RCMF  28.0        700000   
2                      Catanzaro     CF, LWF, AMF  28.0        550000   
3               Sangiuliano City     CF, RWF, AMF  30.0        500000   
4                      Catanzaro     CF, AMF, LWF  33.0        500000   

  Contract expires  Matches played  Minutes played  ...  \
0              NaN              38            3109  ...   
1       2024-06-30              36            3341  ...   
2       2024-06-30              38            2488  ...   
3       2024-06-30              34            2639  ...   
4       2024-06-30              38            1456  ...   

   Prevented goals per 90  Back passes received as GK per 90  Exits per 90  \
0                     NaN                               0.00           NaN   
1                     NaN                               0.32           NaN   
2                     NaN                               0.07           NaN   
3                     NaN                               0.10           NaN   
4                     NaN                               0.00           NaN   

   Aerial duels per 90.1  Free kicks per 90  Direct free kicks per 90  \
0                    NaN               0.00                      0.00   
1                    NaN               1.29                      0.11   
2                    NaN               0.00                      0.00   
3                    NaN               2.28                      0.55   
4                    NaN               0.62                      0.31   

  Direct free kicks on target, % Corners per 90 Penalties taken  \
0                            0.0           0.00               5   
1                           25.0           1.97               0   
2                            0.0           0.00               2   
3                           50.0           3.04               0   
4                           20.0           0.31               1   

   Penalty conversion, %  
0                   80.0  
1                    0.0  
2                   50.0  
3                    0.0  
4                    0.0  

[5 rows x 115 columns]

In [92]:
# Extracting features and target
X = NAC_df['Age'].values
y = NAC_df['Market value'].values
X = X.reshape(-1, 1)
# Impute NaN values with the mean
imputer = SimpleImputer(strategy='mean')
X_imputed = imputer.fit_transform(X)

In [93]:
# Installing K Nearest Neighbor Classifier model
knn = KNeighborsClassifier(n_neighbors=7)
knn.fit(X_imputed,y)

# Splitting the data
X_train, X_test, y_train, y_test = train_test_split(X_imputed, y, test_size=0.4, random_state=42)
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)

In [94]:
# Generating the confusion matrix and classification report
print(confusion_matrix(y_test, y_pred))

[[596   0   2 ...   0   0   0]
 [ 14   0   3 ...   0   0   0]
 [ 36   0   1 ...   0   0   0]
 ...
 [  1   0   0 ...   0   0   0]
 [  0   0   0 ...   0   0   0]
 [  0   0   0 ...   0   0   0]]


In [95]:
# Calculated f1-score and precision sepparately as the function encountered multiple 'zero-division' instances
precision = precision_score(y_test, y_pred, average='weighted', zero_division=1)
f1 = f1_score(y_test, y_pred, average='weighted', zero_division=1)
report = classification_report(y_test, y_pred)
print("Precision:", precision)
print("F1-Score:", f1)
print("Classification Report:")
print(report)

Precision: 0.5231400550690446
F1-Score: 0.10087649101979862
Classification Report:
              precision    recall  f1-score   support

           0       0.34      0.55      0.42      1092
       10000       0.00      0.00      0.00        28
       25000       0.02      0.01      0.01        70
       50000       0.02      0.06      0.03       198
       75000       0.03      0.16      0.04       182
      100000       0.00      0.00      0.00       341
      125000       0.00      0.00      0.00       139
      150000       0.07      0.11      0.08       453
      175000       0.00      0.00      0.00       178
      200000       0.13      0.20      0.16       566
      225000       0.00      0.00      0.00        84
      250000       0.02      0.02      0.02       420
      275000       0.00      0.00      0.00        59
      300000       0.09      0.13      0.11       462
      325000       0.00      0.00      0.00        36
      350000       0.00      0.00      0.00       23

/Users/daria/anaconda3/envs/block_b/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/daria/anaconda3/envs/block_b/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/daria/anaconda3/envs/block_b/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(r

In [96]:
# Import LogisticRegression
from sklearn.linear_model import LogisticRegression

# Instantiate the model
logreg = LogisticRegression()

# Fit the model
logreg.fit(X_train, y_train)

# Predict probabilities
y_pred_probs = logreg.predict_proba(X_test)[:, 1]

print(y_pred_probs[:10])

[0.00479806 0.00552442 0.00436481 0.00560511 0.00542846 0.00223915
 0.00564496 0.00567587 0.0038607  0.00297659]


/Users/daria/anaconda3/envs/block_b/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


## Breast Cancer Wisconsin Data

### 1. Data Preprocessing

In [97]:
wisconsin_data = pd.read_excel('Wisconsin_BreastCancer_data.xlsx')
breast_cancer_df = pd.DataFrame(wisconsin_data)
print(breast_cancer_df.head())

         id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_wor

Checking for missing values

In [98]:
if breast_cancer_df.isnull().any().any():
    print('There are missing values.')
else:
    print("There are no missing values in the Dataset")
missing_values = breast_cancer_df.isnull().sum()

There are no missing values in the Dataset


Converting categorical data into numerical

In [99]:
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()
breast_cancer_df['diagnosis'] = label_encoder.fit_transform(breast_cancer_df['diagnosis'])
breast_cancer_df

id  diagnosis  radius_mean  texture_mean  perimeter_mean  \
0      842302          1        17.99         10.38          122.80   
1      842517          1        20.57         17.77          132.90   
2    84300903          1        19.69         21.25          130.00   
3    84348301          1        11.42         20.38           77.58   
4    84358402          1        20.29         14.34          135.10   
..        ...        ...          ...           ...             ...   
564    926424          1        21.56         22.39          142.00   
565    926682          1        20.13         28.25          131.20   
566    926954          1        16.60         28.08          108.30   
567    927241          1        20.60         29.33          140.10   
568     92751          0         7.76         24.54           47.92   

     area_mean  smoothness_mean  compactness_mean  concavity_mean  \
0       1001.0          0.11840           0.27760         0.30010   
1       1326.0          0.08474           0.07864         0.08690   
2       1203.0          0.10960           0.15990         0.19740   
3        386.1          0.14250           0.28390         0.24140   
4       1297.0          0.10030           0.13280         0.19800   
..         ...              ...               ...             ...   
564     1479.0          0.11100           0.11590         0.24390   
565     1261.0          0.09780           0.10340         0.14400   
566      858.1          0.08455           0.10230         0.09251   
567     1265.0          0.11780           0.27700         0.35140   
568      181.0          0.05263           0.04362         0.00000   

     concave points_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0                0.14710  ...        25.380          17.33           184.60   
1                0.07017  ...        24.990          23.41           158.80   
2                0.12790  ...        23.570          25.53           152.50   
3                0.10520  ...        14.910          26.50            98.87   
4                0.10430  ...        22.540          16.67           152.20   
..                   ...  ...           ...            ...              ...   
564              0.13890  ...        25.450          26.40           166.10   
565              0.09791  ...        23.690          38.25           155.00   
566              0.05302  ...        18.980          34.12           126.70   
567              0.15200  ...        25.740          39.42           184.60   
568              0.00000  ...         9.456          30.37            59.16   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0        2019.0           0.16220            0.66560           0.7119   
1        1956.0           0.12380            0.18660           0.2416   
2        1709.0           0.14440            0.42450           0.4504   
3         567.7           0.20980            0.86630           0.6869   
4        1575.0           0.13740            0.20500           0.4000   
..          ...               ...                ...              ...   
564      2027.0           0.14100            0.21130           0.4107   
565      1731.0           0.11660            0.19220           0.3215   
566      1124.0           0.11390            0.30940           0.3403   
567      1821.0           0.16500            0.86810           0.9387   
568       268.6           0.08996            0.06444           0.0000   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
0                  0.2654          0.4601                  0.11890  
1                  0.1860          0.2750                  0.08902  
2                  0.2430          0.3613                  0.08758  
3                  0.2575          0.6638                  0.17300  
4                  0.1625          0.2364                  0.07678  
..                    ...             ...                      ...  
564                0

Split the dataset into features (X) and target variable (y)

In [100]:
X_bc = breast_cancer_df.drop(["diagnosis", 'id'], axis=1).values
y_bc = breast_cancer_df['diagnosis'].values

Split the data into training and testing sets.

In [101]:
X_bc_train, X_bc_test, y_bc_train, y_bc_test = train_test_split(X_bc, y_bc, test_size=0.2, random_state=42)

# Display the missing values and the shape of the dataset
missing_values, X_bc_train.shape, X_bc_test.shape, y_bc_train.shape, y_bc_test.shape

(id                         0
 diagnosis                  0
 radius_mean                0
 texture_mean               0
 perimeter_mean             0
 area_mean                  0
 smoothness_mean            0
 compactness_mean           0
 concavity_mean             0
 concave points_mean        0
 symmetry_mean              0
 fractal_dimension_mean     0
 radius_se                  0
 texture_se                 0
 perimeter_se               0
 area_se                    0
 smoothness_se              0
 compactness_se             0
 concavity_se               0
 concave points_se          0
 symmetry_se                0
 fractal_dimension_se       0
 radius_worst               0
 texture_worst              0
 perimeter_worst            0
 area_worst                 0
 smoothness_worst           0
 compactness_worst          0
 concavity_worst            0
 concave points_worst       0
 symmetry_worst             0
 fractal_dimension_worst    0
 dtype: int64,
 (455, 30),
 (114, 30),
 

### 2. Model Training

Import the logistic regression model from a library like scikit-learn.

In [102]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

Train the model on the training set

In [103]:
# Instantiate the model
bc_logreg = LogisticRegression(max_iter=10000)

# Fit the model
bc_logreg.fit(X_bc_train, y_bc_train)

LogisticRegression(max_iter=10000)

### 3. Model Evaluation

Predict the target variable for the test set.

In [104]:
y_bc_pred = bc_logreg.predict(X_bc_test)

Evaluate the model

In [108]:
accuracy = accuracy_score(y_bc_test, y_bc_pred)
conf_matrix = confusion_matrix(y_bc_test, y_bc_pred)
class_report = classification_report(y_bc_test, y_bc_pred)
print(accuracy)
print(conf_matrix)
print(class_report)

0.956140350877193
[[70  1]
 [ 4 39]]
              precision    recall  f1-score   support

           0       0.95      0.99      0.97        71
           1       0.97      0.91      0.94        43

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114



The logistic regression model has been successfully trained and evaluated. Here are the Evaluation metrics:

#### Accuracy: 
The model achieved an accuracy of approximately 95.61%. This indicates a high level of correct predictions.


#### Confusion Matrix:

True negatives: 70 (actual negatives correctly predicted)

False positives: 1 (actual negatives incorrectly predicted as positive)

False negatives: 4 (actual positives incorrectly predicted as negative)

True positives: 39 (actual positives correctly predicted)


#### Classification Report:

Precision for class 0 (negative) is 0.95 and for class 1 (positive) is 0.97. Precision is the ratio of correctly predicted positive observations to the total predicted positives.

Recall for class 0 is 0.99 and for class 1 is 0.91. Recall is the ratio of correctly predicted positive observations to all observations in the actual class.

The F1-score is a weighted average of Precision and Recall. It is 0.97 for class 0 and 0.94 for class 1.

The report provides these metrics for both classes, along with macro and weighted averages.